### DISCUSSION TITLE
setup OpenTelemetry with AWS Lambda and chain to parent traceId

### DISCUSSION THREAD

**Discussion by jimkont** [quarkusio/quarkus#31552] [Q&A]
I set quarkus with opentelemetry and AWS lambda as described in the respective documentation pages, however, it seems that I am missing some parts of the process. 

while the traces are correctly propagated to jaeger, they are not chained with the parent traceId that is passed in the lambda invocation. A similar node-based lambda function we have does that automatically so not sure if there is any extra configuration we need to perform in quarkus to pickup the traceId from the lambda invocation context.
Does anyone have any experience in this type of setup?

[a comment by an automated account is not shown]

**Comment by brunobat** [comment]
What propagation mechanism are you assuming? AWS default context propagation uses xray... Are you including that propagator on the Quarkus project?

**Reply by jimkont** [reply]
Let me explain in a bit more detail what the problem is. 

We have a central service that calls multiple lambda functions. That service has opentelemetry enabled and passes a traceId along with every AWS Lambda invocation. An existing lambda function implemented in nodeJs automatically picks up the traceId from the central service and chains it with the traces it produces.
Our quarkus-based lambda function generates traces correctly but does not chain the traces with the traceId that is passed from the central service. The problem with this is that when we trace a central request, the traces from our quarkus-lambda are not shown there, although they do exist separately.

I guess that there is some extra configuration needed to pick it up or maybe something related to https://github.com/quarkusio/quarkus/issues/30926.

I am new to OTel and not 100% familiar with the terminology so apologies in advance for that

**Reply by brunobat** [reply]
What type of headers format are you using in the central service for propagation? W3C, B3 or x-ray?
Knowing that will help a lot.

**Reply by jimkont** [reply]
It should be W3C (not X-Ray, neither B3)

**Reply by brunobat** [reply]
Are you able to check if that header arrives to the lambda?

**Reply by jimkont** [reply]
I could not find a way to check the headers. Is there a way I can access them behind Quarkus?
Besides that, the lambda context object contained no trace-related information. I could also access the `_X_AMZN_TRACE_ID` environment variable that contained a value like `Root=1-640643e4-111bef2e0567ca2f47d34d3d;Parent=0fe0ad42010c549a;Sampled=0`. However, the trace there did not match what the central service was sending

**Reply by brunobat** [reply]
According to OTel, you need to include the X-Ray propagator.
Please check: https://opentelemetry.io/docs/reference/specification/trace/semantic_conventions/instrumentation/aws-lambda/#aws-x-ray-environment-span-link

**Reply by brunobat** [reply]
Also check the Quarkus documentation on how to add that propagation, here: https://quarkus.io/guides/opentelemetry#propagators

**Reply by jimkont** [reply]
update: in the lambda request that is sent by the central service, I see this header:
```
x-amz-client-context: eyJDdXN0b20iOnsidHJhY2VwYXJlbnQiOiIwMC1kZWQ4ZGUwMzNjYWRlMjY3MThkM2Y5ZjUwMzUwZDg0ZS0xMWU0OTllYTI4NWFkODIwLTAxIn19
```
which is a base64 encoding of `{"Custom":{"traceparent":"00-ded8de033cade26718d3f9f50350d84e-11e499ea285ad820-01"}}`

however, when I run the configuration locally by submitting the call to the dev endpoint (`http://localhost:8080`) as well as online, the AWS lambda `Context.clientContext` object is null, is there another way I can access the value of `x-amz-client-context` header here?

**Reply by brunobat** [reply]
Please check my last comments. You MUST use X-Ray propagation with AWS Lambdas and therefore include the related libs in the Quarkus service, otherwise, nothing will work.

**Reply by jimkont** [reply]
Thanks for the links, looking into it already. 

However I need some clarification, the [AWS lambda invocation docs](https://docs.aws.amazon.com/lambda/latest/dg/API_Invoke.html) states that `X-Amz-Client-Context` can be used to populate the context.clientObject with client data, however, I see that in the [quarkus lambda library](https://github.com/quarkusio/quarkus/blob/main/extensions/amazon-lambda/common-runtime/src/main/java/io/quarkus/amazon/lambda/runtime/AmazonLambdaContext.java#L52-L55) you ignore that header when you populate the context object and only read `Lambda-Runtime-Client-Context`. Is there another place that you read this information or does quarkus discard it completely?

**Comment by brunobat** [comment]
Because you need the X-Ray propagator included in 

```
<dependency>
    <groupId>io.opentelemetry</groupId>
    <artifactId>opentelemetry-extension-aws</artifactId>
</dependency>
```

For Quarkus to do that.

**Reply by jimkont** [reply]
Many apologies for persisting but I think I am confused. 

From my understanding, `io.opentelemetry:opentelemetry-extension-aws` is set to read the `X-Amzn-Trace-Id` header, which is required for X-Ray. I have tried to use that library without success.

In my previous message, I was referring to the `X-Amz-Client-Context` header. This is a [standard AWS Lambda header](https://docs.aws.amazon.com/lambda/latest/dg/API_Invoke.html) used to pass client metadata to the handler and this is how our central service is passing the tracing information to all lambda functions. From what I understand this header has nothing to do with OpenTelemetry.
The [Quarkus AmazonLambdaContext](https://github.com/quarkusio/quarkus/blob/main/extensions/amazon-lambda/common-runtime/src/main/java/io/quarkus/amazon/lambda/runtime/AmazonLambdaContext.java#L52-L55) class seems to be ignoring this header when building the handler context and I think this is where the issue in my case lies.

**Reply by jimkont** [reply]
created this issue as a follow-up https://github.com/quarkusio/quarkus/issues/31676, 
thanks for your help so far @brunobat
